docs/proposals/nodeselector.md

```diff
@@ -0,0 +1,174 @@
+we will *add* the `NodeConstraints` field rather than replacing `map[string]string`,
+due to backward compatibility requirements.) We further propose an additional
+new field for `PodSpec`, called `NodePreferences`, that specifies which nodes are
+preferred for scheduling among those that meet the hard requirements described
+by `NodeConstraints`.
+
+`NodeConstraints` and `NodePreferences` allow a pod to request various properties
+that are inherent to nodes, for example "run this pod on a node with an Intel CPU" or, in a
+multi-zone cluster, "run this pod on a node in zone Z."
+([This issue](https://github.com/kubernetes/kubernetes/issues/9044) describes
+some of the properties that a node might publish as labels, which `NodeConstraints`
+and `NodePreferences` can match against.)
+They do *not* allow a pod to request to schedule
+(or not schedule) on a node based on what other pods are running on the node. That
+feature is called "inter-pod topological affinity/anti-afinity" and is described
+[here](https://github.com/kubernetes/kubernetes/pull/18265).
+
+## API
+
+### NodeSelector
+
+```go
+// A node selector represents the union of the results of one or more label queries
+// over a set of nodes; that is, it represents the OR of the selectors represented
+// by the nodeSelectorTerms.
+// TODO: what are the semantics of empty and null?
+type NodeSelector struct {
+	// nodeSelectorTerms is a list of node selector terms. The terms are ORed.
+	NodeSelectorTerms []NodeSelectorTerm `json:"nodeSelectorTerms,omitempty"`
+}
+
+// An empty node selector term matches all objects. A null node selector term
+// matches no objects.
+type NodeSelectorTerm struct {
+	// matchExpressions is a list of node selector requirements. The requirements are ANDed.
+	MatchExpressions []NodeSelectorRequirement `json:"matchExpressions,omitempty"`
+}
+
+// A node selector requirement is a selector that contains values, a key, and an operator
+// that relates the key and values.
+type NodeSelectorRequirement struct {
+	// key is the label key that the selector applies to.
+	Key string `json:"key" patchStrategy:"merge" patchMergeKey:"key"`
+	// operator represents a key's relationship to a set of values.
+	// Valid operators ard In, NotIn, Exists, DoesNotExist. Gt, and Lt.
+	Operator NodeSelectorOperator `json:"operator"`
+	// values is an array of string values. If the operator is In or NotIn,
+	// the values array must be non-empty. If the operator is Exists or DoesNotExist,
+	// the values array must be empty. If the operator is Gt or Lt, the values
+	// array must have a single element, which will be interpreted as an integer.
+    // This array is replaced during a strategic merge patch.
+	Values []string `json:"values,omitempty"`
+}
+
+// A node selector operator is the set of operators that can be used in
+// a node selector requirement.
+type NodeSelectorOperator string
+
+const (
+	NodeSelectorOpIn           NodeSelectorOperator = "In"
```

Comment:
We should also document what scheduling throughput we want at the maximum cluster size we expect to support. At least hundreds of pods per second, I'd guess.

Also, good luck explaining to the user why pods with complex constraint expressions don't schedule.

Use cases that need something more complex than this would likely be better served by something other than constraint expressions.